File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int max = 0;
        int i = 0;
        int j = heights.length - 1;
        while (i < j) {
            if (heights[i] < heights[j]) {
                max = Math.max(max, heights[i] * (j - i));
                i++;
            } else {
                max = Math.max(max, heights[j] * (j - i));
                j--;
            }
        }
        return max;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-0.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        minProduct = nums[0]
        maxProduct = nums[0]
        res = nums[0]

        for i in range(1, len(nums)):
            num = nums[i]
            if num < 0:
                minProduct, maxProduct = maxProduct, minProduct
            minProduct = min(minProduct * num, num)
            maxProduct = max(maxProduct * num, num)

            res = max(res, maxProduct)

        return res